bundles/mapping/shadow-plugin-3d/plugin/ShadowingPluginClass.js

```diff
@@ -0,0 +1,188 @@
+ */
+class ShadowingPlugin extends BasicMapModulePlugin {
+    constructor (instance) {
+        super();
+        this.instance = instance;
+        this._clazz = className;
+        this._name = 'ShadowingPlugin';
+        this._defaultLocation = 'top right';
+        this._time = null;
+        this._date = null;
+        this._log = Oskari.log(className);
+        this.loc = Oskari.getMsg.bind(null, 'ShadowingPlugin3d');
+        this._toolOpen = false;
+        this._element = null;
+        this._popupContent = null;
+        this._popup = null;
+        this._mountPoint = jQuery('<div class="mapplugin shadow-plugin"><div></div></div>');
+        this._popupTemplate = jQuery('<div></div>');
+        this.stateHandler = new ShadowToolHandler(this.instance);
+    }
+    getName () {
+        return className;
+    }
+
+    /**
+     * @method resetState
+     * Resets the state in the plugin
+     */
+    resetState () {
+        this.redrawUI(Oskari.util.isMobile());
+    }
+
+    redrawUI (mapInMobileMode) {
+        this.teardownUI();
+        this._createUI(mapInMobileMode);
+    }
+
+    _createEventHandlers () {
+        return {
+            TimeChangedEvent: function (event) {
+                this._time = event.getTime();
+                this._date = event.getDate();
+            }
+        };
+    }
+
+    getElement () {
+        return this._element;
+    }
+
+    getPopUp () {
+        return this._popup;
+    }
+
+    _teardownUI () {
+        if (!this.getElement()) {
+            return;
+        }
+        ReactDOM.unmountComponentAtNode(this.getElement().get(0));
+        this.getElement().detach();
+    }
+
+    stopPlugin () {
+        this._teardownUI();
+    }
+
+    _createUI (mapInMobileMode) {
+        this._element = this._mountPoint.clone();
+
+        if (mapInMobileMode) {
+            this._element.css('display', 'inline-block');
+            this._addToMobileToolBar();
+        } else {
+            this._addToPluginContainer();
+        }
+
+        this._createControlElement();
+
+        ReactDOM.render(
+            <LocaleContext.Provider value={this.loc}>
+                <ShadowControl mapInMobileMode={mapInMobileMode}/>
+            </LocaleContext.Provider>, this._element.get(0));
+    }
+
+    _createControlElement () {
+        const me = this;
+        const el = this.getElement();
+
+        el.off('click');
+        el.on('click', function (event) {
+            me._toggleToolState();
+            event.stopPropagation();
+        });
+    }
+
+    _addToMobileToolBar () {
+        const resetMapStateControl = jQuery('.toolbar_mobileToolbar').find('.mobile-reset-map-state');
+        jQuery(this._element).insertAfter(resetMapStateControl);
+    }
+
+    _addToPluginContainer () {
+        const panButtonsControl = jQuery('.mappluginsContent').find('.panbuttons');
+        jQuery(this._element).insertAfter(panButtonsControl);
+    }
```

Comment:
Use "index" to assign "priority"/placement among other plugins: https://github.com/oskariorg/oskari-frontend/blob/master/bundles/framework/coordinatetool/plugin/CoordinateToolPlugin.js#L29